design/src/rfcs/rfc0041_env_defined_service_config.md

```diff
@@ -0,0 +1,242 @@
+-----------
+
+* **Global configuration**: configuration which will be used for requests to any service. May be overridden by service-specific configuration.
+* **Service-specific configuration**: configuration which will be used for requests only to a specific service.
+* **Configuration variable**: A key-value pair that defines configuration e.g. `key = value`, `key: value`, `KEY=VALUE`, etc.
+    * Key and value as used in this RFC refer to each half of a configuration variable.
+* **Sub-properties**: When parsing config variables from a profile file, sub-properties are a newline-delimited list of key-value pairs in an indented block following a `<service name>=\n` line. _For an example, see the **Profile File Configuration** section of this RFC where sub-properties are declared for two different services._
+
+The user experience if this RFC is implemented
+----------------------------------------------
+
+While users can already set global configuration in their environment, this RFC proposes two new ways to set service-specific configuration in their environment.
+
+### Environment Variables
+
+When defining service-specific configuration with an environment variable, all keys are formatted like so:
+
+```sh
+"AWS" + "_" + "<config key in CONST_CASE>" + "_" + "<service ID in CONST_CASE>"
+```
+
+As an example, setting an endpoint URL for different services would look like this:
+
+```sh
+export AWS_ENDPOINT_URL=http://localhost:4567
+export AWS_ENDPOINT_URL_ELASTIC_BEANSTALK=http://localhost:5567
+export AWS_ENDPOINT_URL_DYNAMODB=http://localhost:5678
+```
+
+The first variable sets a global endpoint URL. The second variable overrides the first variable, but only for the Elastic Beanstalk service. The third variable overrides the first variable, but only for the DynamoDB service.
+
+### Profile File Configuration
+
+When defining service-specific configuration in a profile file, it looks like this:
+
+```
+[profile dev]
+services = testing-s3-and-eb
+endpoint_url = http://localhost:9000
+
+[services testing-s3-and-eb]
+s3 =
+  endpoint_url = http://localhost:4567
+elastic_beanstalk =
+  endpoint_url = http://localhost:8000
+```
+
+When `dev` is the active profile, all services will use the `http://localhost:9000` endpoint URL except where it is overridden. Because the `dev` profile references the `testing-s3-and-eb` services, and because two service-specific endpoint URLs are set, those URLs will override the `http://localhost:9000` endpoint URL when making requests to S3 (`http://localhost:4567`) and Elastic Beanstalk (`http://localhost:8000`).
+
+### Configuration Precedence
+
+When configuration is set in multiple places, the value used is determined in this order of precedence:
+
+*highest precedence*
+
+1. Programmatic client configuration
+2. Service-specific environment variables
+3. Global environment variables
+4. Service-specific profile file variables in the active profile
+5. Global profile file variables in the active profile
```

Comment:
While sort-of touched upon already, it'd be helpful to clarify again within this precedence list which ones already exist and which ones are to be implemented.